File: FE/src/app/services/verificationService.ts
import api from './api';

export type VerificationStatus = 'PENDING' | 'APPROVED' | 'REJECTED';

export interface VerificationItem {
  id: string;
  userId: string;
  userName?: string;
  avatarUrl?: string;
  course: string;
  claimedGrade: string;
  transcriptUrl?: string;
  status: VerificationStatus;
  reviewedBy?: string;
  reviewedAt?: string;
  note?: string;
  createdAt: string;
}

export interface SubmitVerificationPayload {
  course: string;
  claimedGrade: 'A' | 'A+';
  transcriptUrl?: string;
}

export async function submitVerification(payload: SubmitVerificationPayload): Promise<VerificationItem> {
  const { data } = await api.post('/api/verifications', payload);
  return data.data as VerificationItem;
}

export async function listMyVerifications(): Promise<VerificationItem[]> {
  const { data } = await api.get('/api/verifications/mine');
  return data.data as VerificationItem[];
}

export async function listVerifications(status?: VerificationStatus): Promise<VerificationItem[]> {
  const { data } = await api.get('/api/verifications', {
    params: status ? { status } : {},
  });
  return data.data as VerificationItem[];
}

export async function decideVerification(
  id: string,
  status: 'APPROVED' | 'REJECTED',
  note?: string
): Promise<VerificationItem> {
  const { data } = await api.post(`/api/verifications/${id}/decision`, { status, note });
  return data.data as VerificationItem;
}

File: BE/src/main/java/com/dynforge/be/service/VerificationService.java
package com.dynforge.be.service;

import com.dynforge.be.exception.BadRequestException;
import com.dynforge.be.exception.ResourceNotFoundException;
import com.dynforge.be.mapper.VerificationMapper;
import com.dynforge.be.model.dto.VerificationDecisionRequest;
import com.dynforge.be.model.dto.VerificationRequestDto;
import com.dynforge.be.model.dto.VerificationResponse;
import com.dynforge.be.model.entity.MentorProfile;
import com.dynforge.be.model.entity.User;
import com.dynforge.be.model.entity.VerificationRequest;
import com.dynforge.be.model.enums.VerificationStatus;
import com.dynforge.be.repository.MentorRepository;
import com.dynforge.be.repository.VerificationRepository;
import lombok.RequiredArgsConstructor;
import org.bson.types.ObjectId;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.List;

@Service
@RequiredArgsConstructor
public class VerificationService {

    private final VerificationRepository verificationRepository;
    private final MentorRepository mentorRepository;
    private final VerificationMapper verificationMapper;

    public VerificationResponse submit(User user, VerificationRequestDto dto) {
        VerificationRequest request = VerificationRequest.builder()
                .userId(new ObjectId(user.getId()))
                .course(dto.course())
                .claimedGrade(dto.claimedGrade())
                .transcriptUrl(dto.transcriptUrl())
                .status(VerificationStatus.PENDING)
                .createdAt(Instant.now())
                .build();

        return verificationMapper.toResponse(verificationRepository.save(request));
    }

    public List<VerificationResponse> getMine(User user) {
        return verificationRepository.findByUserId(new ObjectId(user.getId()))
                .stream().map(verificationMapper::toResponse).toList();
    }

    public List<VerificationResponse> list(VerificationStatus status) {
        List<VerificationRequest> requests = status == null
                ? verificationRepository.findAll()
                : verificationRepository.findByStatus(status);

        return requests.stream().map(verificationMapper::toResponse).toList();
    }

    public VerificationResponse decide(User reviewer, String id, VerificationDecisionRequest decision) {
        if (decision.status() == VerificationStatus.PENDING) {
            throw new BadRequestException("Decision must be APPROVED or REJECTED");
        }

        VerificationRequest request = verificationRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Verification request not found: " + id));

        request.setStatus(decision.status());
        request.setNote(decision.note());
        request.setReviewedBy(new ObjectId(reviewer.getId()));
        request.setReviewedAt(Instant.now());
        request = verificationRepository.save(request);

        if (decision.status() == VerificationStatus.APPROVED) {
            // Upsert: create a minimal profile if the mentor hasn't set one up yet,
            // so approval always results in a verified, listable mentor.
            ObjectId mentorUserId = request.getUserId();
            MentorProfile profile = mentorRepository.findByUserId(mentorUserId)
                    .orElseGet(() -> MentorProfile.builder().userId(mentorUserId).build());
            profile.setVerified(true);
            mentorRepository.save(profile);
        }

        return verificationMapper.toResponse(request);
    }
}
